### DISCUSSION TITLE
Is there a specific order in which transforms are composed?

### DISCUSSION THREAD

**Discussion by sarthakpati** [TorchIO-project/torchio#600] [Q&A] [answered]
So, I am adding multiple transforms together:
- Normalize
- Resample
- Bias
- Noise
- Blur

Does the `torchio.transforms.Compose` ensure a particular order or does it always take the order I have passed? 

The reason why I am asking is because I would imaging certain augmentations to work on non-normalized images (for example, ghosting and spike)...

Cheers,
Sarthak

**Comment by romainVala** [comment]
Hi
the order is the one you passed.
You are right the order matter, but it is let to the user responsibility, since it will be greatly dependent on the application (on the data too).

For instance, Bias and Noise or Noise and Bias. If one which to stay as close as possible to the real acquisition process, one should choose Bias first then noise (because during MRI acquisition the thermal noise, is added uniformly in space to the data.

**Reply by sarthakpati** [reply]
So, if the following are selected for MRI input:
- Normalize
- Resample
- Motion
- Ghosting
- Spike
- Blur
- Noise
- Gamma
- Affine

Would the correct order be the following?

Bias > Motion > Ghosting > Spike > Noise > Gamma > Normalize > Resample > Blur > Affine

**Comment by romainVala** [comment, marked as answer]
this is a difficult question, and I do not have the correct answer, (especially to know if it really matter or not).
So base essentially on my intuition on the MRI acquisition process, I would say

Affine / Resample first because it is to change to subject shape (and it reduce the noise due to interpolation)

Blur  (not sure why to add blur at all, if it is to simulate different acquisition voxel size, I would prefer a random resampling (but it is not implemented ...)

Gamma   (to change the subject intrinsic contrast, before applying other artefact)

bias

Ghosting /  Spike  (spike are quite rare in practice and the current implementation make some strong artefact, ... so to be use with caution)

Noise (at the end so that the level of noise is really the one you put   (plus the one present on your original image)

Normalize (ie rescale or z_normalisation) 

Normalize if it is done before, you may not end up with the desire control range (and  when you do inference you usually do not add artefat, but just normalize the intensity)

I hope it helps

**Reply by sarthakpati** [reply]
Understood, thank you!

**Reply by fepegar** [reply]
> Blur (not sure why to add blur at all, if it is to simulate different acquisition voxel size, I would prefer a random resampling (but it is not implemented ...)

Do you mean [`RandomAnisotropy`](https://torchio.readthedocs.io/transforms/augmentation.html#randomanisotropy)?

**Reply by romainVala** [reply]
Ah yes ! I did not notice this was implemented.  nice !

In the original billot paper, they say : 
` We simulate voxel thickness independently for each channel of I, by blurring them with anisotropic
Gaussian kernels to simulate the target resolution of the LR images. ` 

but in  [RandomAnisotropy](https://torchio.readthedocs.io/transforms/augmentation.html#randomanisotropy)  the implementation is different, since it rely on 2 resampling (down, and up to go back ...). Probably slower to perform to resampling than one bluring ... Although they claim it is equivalent, resampling seems closer to the acquisition process ... 

we are a bit of subject here, but why fixing 'nearest' interpolation for downsampling, I would prefer to go with linear, 
for instance if you downsample from 1mm resolution to 2mm then you have 8 1mm voxels what is the nearest interpolation exactly doing (what is the nearest voxel then ?)

The best choice is to take the mean of those 8 voxels, ( os I think linear would be the best approximation ...)

**Reply by fepegar** [reply]
[This is quite off-topic]

I actually implemented that before reading the paper.

> Although they claim it is equivalent, resampling seems closer to the acquisition process

I agree. The rationale for choosing nearest for downsampling is to have an image that actually looks anisotropic and then upsample it with any other interpolation chosen by the user, which seems natural. But maybe you're right. Let's try:

```python
import torchio
import torchio as tio
t1 = tio.datasets.Colin27().t1
down_nn = tio.Resample((1, 1, 8), image_interpolation='nearest')(t1)
down_lin = tio.Resample((1, 1, 8), image_interpolation='linear')(t1)
up_nn = tio.Resample()(down_nn)
up_lin = tio.Resample()(down_lin)
t1.plot()
up_nn.plot()
up_lin.plot()
```

![Figure_1](https://user-images.githubusercontent.com/12688084/125449100-021a0d9d-f99d-43af-a8bf-4905fb59f4fe.png)
![Figure_2](https://user-images.githubusercontent.com/12688084/125449103-404eadd5-1437-4ed5-9621-4349357a597a.png)
![Figure_3](https://user-images.githubusercontent.com/12688084/125449104-bc5d9b96-efef-45df-8386-329488b9c005.png)

I guess there's not a big difference in practice.

**Reply by romainVala** [reply]
a difference that is not visible by eyes can lead to big difference by CNN ... so one never know for sure ...
I guess it is relative ...

```python
diff_img = up_nn.data - up_lin.data
diff_img.max()
Out[15]: tensor(1742030.7500)

diff_img.max()/t1.data.max()
Out[21]: tensor(0.1806)

t1.set_data(diff_img)
t1.plot()
```
![Figure_1](https://user-images.githubusercontent.com/5611962/125454012-bfa6d810-2e28-43d6-b5db-ffdec97a2e2c.png)

**Reply by sarthakpati** [reply]
Loving this discussion!

Is there any way to make this a bit user-friendly? Basically, if multiple of these transforms are passed (especially the ones for which normalization should **not** be done), can that be handled during transform composition itself?

**Reply by fepegar** [reply]
@romainVala Good point! At the moment I can't look into it, but if you think it's important we can investigate further.

@sarthakpati I think that wouldn't be straightforward. I think the most important thing is to visually inspect some of your augmented images and make sure they look as expected. Some things I try to be careful with:
- If I use default parameters for noise, it should happen after normalization
- If I use noise, don't blur after that

**Reply by romainVala** [reply]
@fepegar not sure it is that important ...

@sarthakpati : Again the exact type of tranformation and the order, will very much depend on the use case. so I do not see any way to do it automatically

What is your use case by the way ?

**Reply by sarthakpati** [reply]
Use case is a DL pipeline for MR images:

```
So, if the following are selected for MRI input:

Normalize
Resample
Motion
Ghosting
Spike
Blur
Noise
Gamma
Affine
```
